### PR TITLE
[SYCL] Do not emit unneeded static initializations in sycl device code

### PR DESCRIPTION
**Pull request by premanandrao**
Signed-off-by: Premanand M Rao <premanand.m.rao@intel.com>

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 1]
```diff
@@ -4687,6 +4687,10 @@ void CodeGenModule::HandleCXXStaticMemberVarInstantiation(VarDecl *VD) {
   if (DK == VarDecl::Definition && VD->hasAttr<DLLImportAttr>())
     return;
 
+  // Emit only those declarations that must be, when on SYCL device.
+  if (LangOpts.SYCLIsDevice && !MustBeEmitted(VD))
```
Comment (erichkeane) [inline clang/lib/CodeGen/CodeGenModule.cpp:4691]: This doesn't seem to be the right place for this.  This is the initialization, not when the variable itself is being emitted. 

Can you backtrace a little to see what is causing the static member to be emitted, and do the test there?  

Also, do we have the same problem with globals (non-members)?  How about magic-statics?

Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: What's happening with this is that the static variable emission (as well as globals and function-local statics) is being throttled using the usual mechanism in CodeGen and AST routines.  However, the initialization is being generated via a call from Sema:HandleCXXStaticMemberVarInstantiation() and the initialization later forces the variable to be emitted as well (since it gets used).

I have moved the check earlier in Sema.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 2]
```diff
@@ -5150,6 +5150,10 @@ void Sema::InstantiateVariableDefinition(SourceLocation PointOfInstantiation,
   if (TSK == TSK_ExplicitSpecialization)
     return;
 
+  if (getLangOpts().SYCLIsDevice && !Var->hasAttr<SYCLKernelAttr>() &&
+      !Var->hasAttr<SYCLDeviceAttr>())
+    return;
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5155]: This stuff doesn't seem right. That is probably why tests are failing.
`sycl_kernel` attribute can be applied only to a function template.
Only `SYCL_EXTERNAL` functions can have `sycl_device` attribute.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Can you point out the place where "initialization later forces the variable to be emitted as well"?

Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: HandleCXXStaticMemberVarInstantiation -> GetAddrOfGlobalVar -> getMangledName

And the code I mentioned above sees that getMangledName was previously called and emits the variable.

Comment (erichkeane) [reply clang/lib/CodeGen/CodeGenModule.cpp]: What calls that handle-static-member-var-inistantiation?  We should probably be suppressing THAT unless necessary.

Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: That is called from Sema: SemaTemplateInstantiateDecl.cpp via PassToConsumeRAII but there isn't enough information there to suppress it.   I tried, but incorrectly; I couldn't figure out what property of Var I could use there.

Comment (erichkeane) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Perhaps we SHOULD be marking these vardecls as 'used-by-sycl' in some manner. If they aren't referenced in a kernel, we probably shouldn't be instantiating them, right?

That way we could just suppress the instantiate-variable-definition in PerformPendingInstantiations.

Comment (erichkeane) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Although, actually... If these are static-data-members, don't we KNOW that they can't be emitted?  We have an error that prevents them from being used, right?  So 'Is static member && SYCL' should be enough.  Either at the beginning of InstantiateVariableDefinition or in performpendinginstantiations, though I'm not sure where would be better.  See 6096 for how to check static-data-member.

That said, do Variable Templates cause us problems here too?

Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: We diagnose non-const static data members.  I will check out the other places you mentioned.

I haven't specifically tested variable templates, will do so.

Comment (erichkeane) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Ah, right.  Does this patch make sure that this gets emitted anyway if it is used then?  I'd hate for us to suppress this here, but have it be used later/before.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp [hunk 3]
```diff
@@ -5271,7 +5271,8 @@ void Sema::InstantiateVariableDefinition(SourceLocation PointOfInstantiation,
       : Consumer(Consumer), Var(Var) { }
 
     ~PassToConsumerRAII() {
-      Consumer.HandleCXXStaticMemberVarInstantiation(Var);
+      if (Var)
```
Comment (erichkeane) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: Is this change necessary? Does HandleCXXStaticMemberVarinstantiation not peroperly handle the null case?

Just surprised that none of the rest of this got there.

Comment (erichkeane) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: Also, It seems that it would be more readable to put the logic from below here, isntead of 'if (var)', just 'f(!SyclIsDevice || !Var->isStaticDataMember())

Comment (premanandrao) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: > Is this change necessary? Does HandleCXXStaticMemberVarinstantiation not peroperly handle the null case?

No it straight up dereferences the incoming VD.  I could add a NULL check there if you want me to.

Comment (premanandrao) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: > Also, It seems that it would be more readable to put the logic from below here, isntead of 'if (var)', just 'f(!SyclIsDevice || !Var->isStaticDataMember())

I couldn't how to access getLangOpts() in the local struct without passing the this pointer from the enclosing member function.

Comment (premanandrao) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: *figure out

Comment (erichkeane) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:5274]: Ah, right, I missread ASTConsumer as ASTcontext.  Just give it a reference to Sema?

Comment (premanandrao) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp]: Done.  Thanks Erich!

**Comment by Fznamznon** [comment]
Looks like this patch breaks pipes. @MrSidims , could you please take a look at the test logs? Have you seen such errors before?

**Comment by MrSidims** [comment]
> Looks like this patch breaks pipes. @MrSidims , could you please take a look at the test logs? Have you seen such errors before?

I'll take a look, we might need to have to do some staggered changes.

**Comment by MrSidims** [comment]
> > Looks like this patch breaks pipes. @MrSidims , could you please take a look at the test logs? Have you seen such errors before?
> 
> I'll take a look, we might need to have to do some staggered changes.

According to log the failure came from FPGA emulator backend's pass, that recognizes SYCL pipes. I'm currently building the debug version of the emulator locally to reproduce, but...
...that looks weird, after applying the patch I see no changes but clang version metadata in LLVM IR module generated for fpga_pipes.cpp test, hence FPGA emulator backend flow for pipes shan't be affected. At the same time the same tests are passing on windows. Will dig more into this when build is complete.

**Comment by premanandrao** [comment]
Thank you @MrSidims , appreciate your help.

**Comment by premanandrao** [comment]
@Fznamznon and @erichkeane, if either of you could review my latest set of changes, I'd appreciate it.

**Comment by MrSidims** [comment]
> Second, @MrSidims , please take a look and make sure you're happy.

Yeah, I happy.

LGTM

**Comment by premanandrao** [comment]
> Please stop force-pushing, it makes reviews really difficult. Second, @MrSidims , please take a look and make sure you're happy.

Sorry.   I will talk offline with you and figure out the BKM when regular push fails.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 4]
```diff
@@ -1276,7 +1276,12 @@ void CodeGenModule::EmitCtorList(CtorList &Fns, const char *GlobalName) {
     ctor.addInt(Int32Ty, I.Priority);
     ctor.add(llvm::ConstantExpr::getBitCast(I.Initializer, CtorPFTy));
     if (I.AssociatedData)
-      ctor.add(llvm::ConstantExpr::getBitCast(I.AssociatedData, VoidPtrTy));
+      if (I.AssociatedData->getType()->getPointerAddressSpace() !=
+          VoidPtrTy->getAddressSpace())
+        ctor.add(
+            llvm::ConstantExpr::getAddrSpaceCast(I.AssociatedData, VoidPtrTy));
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CodeGenModule.cpp:1282]: You are adding a new address space cast, but I don't see any address space casts in the test.

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 5]
```diff
@@ -1276,7 +1276,12 @@ void CodeGenModule::EmitCtorList(CtorList &Fns, const char *GlobalName) {
     ctor.addInt(Int32Ty, I.Priority);
     ctor.add(llvm::ConstantExpr::getBitCast(I.Initializer, CtorPFTy));
     if (I.AssociatedData)
-      ctor.add(llvm::ConstantExpr::getBitCast(I.AssociatedData, VoidPtrTy));
+      if (I.AssociatedData->getType()->getPointerAddressSpace() !=
+          VoidPtrTy->getAddressSpace())
```
Comment (Fznamznon) [inline clang/lib/CodeGen/CodeGenModule.cpp:1280]: Should we check that it is correct to do address space cast here?
Like this check
https://github.com/intel/llvm/blob/ccefc93c884556ca1de01cfe2184274ea7c1d2f7/clang/lib/CodeGen/CGExprScalar.cpp#L2996
@asavonic ?

File: clang/test/CodeGenSYCL/sycl-device-static-init.cpp [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+template <class T>
+const RegisterBaseInit BaseInit<T>::s_regbase;
+template <class T>
+RegisterBaseInit BaseInit<T>::s_regbase_ncsdm;
+template <class T>
+const int BaseInit<T>::var = 9;
+template struct BaseInit<TestBaseType>;
+template <typename name, typename Func>
+__attribute__((sycl_kernel)) void kernel_single_task(Func kernelFunc) {
+  kernelFunc(BaseInit<TestBaseType>::s_regbase);
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/sycl-device-static-init.cpp:35]: I'm not able to understand what is happening in this test anymore?
Could you please elaborate?

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 5 above)
Comment (asavonic) [reply clang/lib/CodeGen/CodeGenModule.cpp:1280]: Yes, we have to ensure that an addrspace cast is valid.

File: clang/test/CodeGenSYCL/sycl-device-static-init.cpp (hunk 6 above)
Comment (premanandrao) [reply clang/test/CodeGenSYCL/sycl-device-static-init.cpp:35]: The test is checking that the non-const static data member is not emitted.  We continue to emit the const members.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 4 above)
Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp:1282]: Thanks, I will modify the test to show this.

File: clang/lib/CodeGen/CodeGenModule.cpp (hunk 5 above)
Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp:1280]: Okay, will add this check.

Comment (premanandrao) [reply clang/lib/CodeGen/CodeGenModule.cpp]: In the cases above, it looks like they values are both cast to the generic address space, to ensure sameness.   This does not seem to be needed here, as the cast is to address space 0.  However, I have changed the call to use an existing routine (which does the check as well) which is called in few other places, instead of checking it directly how I had it before.

File: clang/test/CodeGenSYCL/sycl-device-static-init.cpp [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+// Test that static initializers do not force the emission of globals on sycl device
+
+// CHECK: %struct._ZTS16RegisterBaseInit.RegisterBaseInit = type { i8 }
+// CHECK-NOT: $_ZN8BaseInitI12TestBaseTypeE15s_regbase_ncsdmE = comdat any
+// CHECK: $_ZN8BaseInitI12TestBaseTypeE3varE = comdat any
+// CHECK: @_ZN8BaseInitI12TestBaseTypeE9s_regbaseE = {{.*}} global %struct._ZTS16RegisterBaseInit.RegisterBaseInit
+// CHECK-NOT: @_ZN8BaseInitI12TestBaseTypeE15s_regbase_ncsdmE = weak_odr addrspace(1) global %struct._ZTS16RegisterBaseInit.RegisterBaseInit zeroinitializer, comdat, align 1
+// CHECK: @_ZN8BaseInitI12TestBaseTypeE3varE = weak_odr addrspace(1) constant i32 9, comdat, align 4
+// CHECK: @llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 65535, void ()* @__cxx_global_var_init, i8* addrspacecast (i8 addrspace(1)* getelementptr inbounds (%struct._ZTS16RegisterBaseInit.RegisterBaseInit, %struct._ZTS16RegisterBaseInit.RegisterBaseInit addrspace(1)* @_ZN8BaseInitI12TestBaseTypeE9s_regbaseE, i32 0, i32 0) to i8*) }]
```
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/sycl-device-static-init.cpp:10]: Just one more question, why we still emit `llvm.global_ctors`, AFAIK this is array of constructors for globals and they are called when module is loaded. https://llvm.org/docs/LangRef.html#id1395
In SYCL, global variables which can be used in device code must be const/constexpr and zero-initialized/const-initialized. I'm not sure that we should emit any constructors for global variables at all.

Comment (premanandrao) [reply clang/test/CodeGenSYCL/sycl-device-static-init.cpp:10]: It is being used for initializing the const static data member of the class.  It seems to be needed.   A version without this broke the fpga pipes in testing a couple days back.
